src/io/benchmark.zig

```diff
@@ -0,0 +1,238 @@
+            completion: *IO.Completion,
+            result: IO.AcceptError!os.socket_t,
+        ) void {
+            assert(self.rx.socket.fd == -1);
+            assert(&self.server.completion == completion);
+            self.rx.socket.fd = result catch |err| std.debug.panic("accept error {}", .{err});
+
+            // Start reading data from the accepted client socket
+            assert(self.rx.transferred == 0);
+            self.do_transfer("rx", .read, 0);
+        }
+
+        fn on_connect(
+            self: *Context,
+            completion: *IO.Completion,
+            result: IO.ConnectError!void,
+        ) void {
+            assert(self.tx.socket.fd != -1);
+            assert(&self.tx.socket.completion == completion);
+
+            // Start sending data to the server's accepted client
+            assert(self.tx.transferred == 0);
+            self.do_transfer("tx", .write, 0);
+        }
+
+        const TransferType = enum {
+            read = 0,
+            write = 1,
+        };
+
+        fn do_transfer(
+            self: *Context,
+            comptime pipe_name: []const u8,
+            comptime transfer_type: TransferType,
+            bytes: usize,
+        ) void {
+            // The type of IO to perform and what type of IO to perform next (after the current one completes).
+            const transfer_info = switch (transfer_type) {
+                .read => .{
+                    .IoError = IO.RecvError,
+                    .io_func = "recv",
+                    .next = TransferType.write,
+                },
+                .write => .{
+                    .IoError = IO.SendError,
+                    .io_func = "send",
+                    .next = TransferType.read,
+                },
+            };
+
+            assert(bytes <= buffer_size);
+            self.transferred += bytes;
+
+            // Check in with the benchmark timer to stop sending/receiving data
+            self.current = self.timer.monotonic();
+            if (!self.isRunning())
+                return;
+
+            // Select which connection (tx or rx) depending on the type of transfer
+            const pipe = &@field(self, pipe_name);
+            pipe.transferred += bytes;
+            assert(pipe.transferred <= pipe.buffer.len);
+
+            // There's still more data to transfer on the connection
+            if (pipe.transferred < pipe.buffer.len) {
+                // Callback which calls this function again when data is transferred.
+                // Effectively loops back above.
+                const on_transfer = struct {
+                    fn on_transfer(
+                        _self: *Context,
+                        completion: *IO.Completion,
+                        result: transfer_info.IoError!usize,
+                    ) void {
+                        const _bytes = result catch |err| {
+                            std.debug.panic("{} error: {}", .{ transfer_info.io_func, err });
```

Comment:
Leaving off the specifier for slices is a compile error.
```suggestion
                            std.debug.panic("{s} error: {}", .{ transfer_info.io_func, err });
```